File: iobroker/heizstab-bridge.py
#!/usr/bin/env python3
# Heizstab Manuell/Auto-Brücke: liest Modus aus ioBroker (simple-api REST) und
# stellt SmartFox per Modbus TCP. Self-healing: gleicht alle 3 s ab, schreibt nur
# bei Abweichung (SmartFox quittiert Writes oft verzögert -> Read-Back zählt).
#   AUTO    -> 40400=0            (Control via Modbus AUS -> SmartFox regelt per PV)
#   MANUELL -> 40400=1, 40401=val (val=1000 -> 100%/EIN, 0 -> AUS; Aout 0.1%-Schritte)
import socket, struct, time, urllib.request

SF_IP = "100.64.90.62"; SF_PORT = 502
API = "http://127.0.0.1:8087/getPlainValue/"
MODE = "0_userdata.0.heizstab.mode"
MANON = "0_userdata.0.heizstab.manual_on"

def get_state(sid):
    try:
        with urllib.request.urlopen(API + sid, timeout=4) as r:
            return r.read().decode().strip().strip('"')
    except Exception:
        return None

def mb_req(pdu, timeout=5):
    mbap = struct.pack('>HHHB', 1, 0, len(pdu) + 1, 1)
    try:
        s = socket.create_connection((SF_IP, SF_PORT), timeout=timeout)
        s.sendall(mbap + pdu); r = s.recv(512); s.close(); return r
    except Exception:
        return None

def read_reg(addr):
    r = mb_req(struct.pack('>BHH', 0x03, addr, 1))
    if not r or len(r) < 11 or (r[7] & 0x80):
        return None
    return struct.unpack('>H', r[9:11])[0]

def write_regs(addr, vals):
    pdu = struct.pack('>BHHB', 0x10, addr, len(vals), len(vals) * 2) + b''.join(struct.pack('>H', v & 0xFFFF) for v in vals)
    mb_req(pdu)  # Quittung kommt oft verspätet -> ignorieren, per Read-Back prüfen

def desired():
    mode = get_state(MODE)
    if mode is None:
        return None
    manon = str(get_state(MANON)).lower() in ('true', '1', 'on')
    if mode == 'manuell':
        return (1, 1000 if manon else 0)
    return (0, 0)

print("[heizstab-bridge] start", flush=True)
last_log = None
while True:
    try:
        tgt = desired()
        if tgt is not None:
            ctrl = read_reg(40400)
            if ctrl is not None:
                mism = (ctrl != tgt[0])
                if tgt[0] == 1:
                    aout = read_reg(40401)
                    if aout is not None and aout != tgt[1]:
                        mism = True
                if mism:
                    write_regs(40400, list(tgt))
                    time.sleep(1.0)
                    nc = read_reg(40400)
                    if last_log != tgt:
                        print("[heizstab-bridge] Ziel %s geschrieben, 40400=%s" % (str(tgt), str(nc)), flush=True)
                        last_log = tgt
    except Exception as e:
        print("[heizstab-bridge] err %s" % e, flush=True)
    time.sleep(3)

File: iobroker/heizstab-bridge.js
// Heizstab Manuell/Auto-Steuerbrücke — eigenständig im ioBroker (ersetzt den
// Python-Dienst heizstab-bridge). Liest 0_userdata.0.heizstab.mode/manual_on
// nativ und stellt SmartFox (100.64.90.62:502) per Modbus TCP. Self-healing:
// gleicht alle 3 s ab + bei jeder State-Änderung, schreibt nur bei Abweichung.
//   AUTO    -> 40400=0            (Control via Modbus AUS -> SmartFox regelt per PV)
//   MANUELL -> 40400=1, 40401=val (1000=100%/EIN, 0=AUS; Aout 0.1%-Schritte)
const net = require('net');
const IP = '100.64.90.62', PORT = 502;
const MODE = '0_userdata.0.heizstab.mode';
const MANON = '0_userdata.0.heizstab.manual_on';

function mbReq(buf) {
  return new Promise((resolve) => {
    const s = new net.Socket(); let done = false, acc = Buffer.alloc(0);
    const fin = (r) => { if (done) return; done = true; try { s.destroy(); } catch (e) {} resolve(r); };
    s.setTimeout(5000);
    s.on('timeout', () => fin(null));
    s.on('error', () => fin(null));
    s.on('data', (d) => { acc = Buffer.concat([acc, d]); if (acc.length >= 9) fin(acc); });
    s.connect(PORT, IP, () => s.write(buf));
  });
}
function frameWrite(addr, vals) {
  const q = vals.length, b = Buffer.alloc(13 + q * 2);
  b.writeUInt16BE(1, 0); b.writeUInt16BE(0, 2); b.writeUInt16BE(7 + q * 2, 4); b.writeUInt8(1, 6);
  b.writeUInt8(0x10, 7); b.writeUInt16BE(addr, 8); b.writeUInt16BE(q, 10); b.writeUInt8(q * 2, 12);
  for (let i = 0; i < q; i++) b.writeUInt16BE(vals[i] & 0xFFFF, 13 + i * 2);
  return b;
}
function frameRead(addr, q) {
  const b = Buffer.alloc(12);
  b.writeUInt16BE(1, 0); b.writeUInt16BE(0, 2); b.writeUInt16BE(6, 4); b.writeUInt8(1, 6);
  b.writeUInt8(0x03, 7); b.writeUInt16BE(addr, 8); b.writeUInt16BE(q, 10);
  return b;
}
async function readReg(addr) {
  const r = await mbReq(frameRead(addr, 1));
  if (!r || r.length < 11 || (r[7] & 0x80)) return null;
  return r.readUInt16BE(9);
}
function desired() {
  const m = getState(MODE);
  if (!m || m.val == null) return null;
  const mo = getState(MANON);
  const manon = !!mo && (mo.val === true || mo.val === 'true' || mo.val === 1);
  if (m.val === 'manuell') return [1, manon ? 1000 : 0];
  return [0, 0];
}
let busy = false, loggedOnce = false;
async function apply() {
  if (busy) return; busy = true;
  try {
    const tgt = desired();
    if (!tgt) return;
    const ctrl = await readReg(40400);
    if (!loggedOnce) { log('[Heizstab-Bridge] aktiv — 40400=' + ctrl + ' Ziel=' + JSON.stringify(tgt)); loggedOnce = true; }
    if (ctrl == null) return;
    let mism = ctrl !== tgt[0];
    if (tgt[0] === 1) { const a = await readReg(40401); if (a != null && a !== tgt[1]) mism = true; }
    if (mism) {
      await mbReq(frameWrite(40400, tgt));
      log('[Heizstab-Bridge] gestellt: 40400=' + tgt[0] + ' 40401=' + tgt[1]);
    }
  } finally { busy = false; }
}
on({ id: MODE, change: 'any' }, apply);
on({ id: MANON, change: 'any' }, apply);
setInterval(apply, 3000);
log('[Heizstab-Bridge] Script gestartet (SmartFox ' + IP + ' Modbus 40400/40401)');
apply();
